staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -121,9 +122,50 @@ func (wm watchersMap) terminateAll(done func(*cacheWatcher)) {
 	}
 }
 
+type timeBucketWatchersMap map[int64][]*cacheWatcher
+
+func (bw timeBucketWatchersMap) addWatcher(w *cacheWatcher, id int64) {
+	watchers, ok := bw[id]
+	if ok {
+		watchers = append(watchers, w)
+	} else {
+		watchers = []*cacheWatcher{w}
+	}
+	bw[id] = watchers
+}
+
+func (bw timeBucketWatchersMap) getBucket(id int64) []*cacheWatcher {
+	if watchers, ok := bw[id]; ok {
+		return watchers
+	}
+	return nil
+}
+
+func (bw timeBucketWatchersMap) popBucket(id int64) []*cacheWatcher {
+	if watchers, ok := bw[id]; ok {
+		delete(bw, id)
+		return watchers
+	}
+	return nil
+}
+
+func (bw timeBucketWatchersMap) size() int {
+	return len(bw)
+}
+
 type indexedWatchers struct {
 	allWatchers   watchersMap
 	valueWatchers map[string]watchersMap
+	// Maintain a time wheel to decide if it's already time to send it or not, rather than
+	// go over all watchers. Keep all watchers within 1 second into one bucket, when the time
+	// comes, we will send bookmark event to the entire bucket.
+	watchersQueue timeBucketWatchersMap
```

Comment:
I was thinking about this more and I think that this doesn't belong to this structure.
Up until now, indexedWatchers contains all existing watchers registered exactly once.

watchersQueue contains pointers to the same watchers (or their subset). The API for operating them is different, etc.
I think that this should be explicitly a separate structure.

[BTW - this part of adding this structure and piping it through the code should be moved to a separate PR].

Reply:
Refactored into a separate structure.